docs/design-v1.md

```diff
@@ -0,0 +1,106 @@
+# Caching Service Design Doc
+
+## Purpose
+
+Provide a generalized cache store that can be used from any other service, allowing users to avoid re-running expensive computation that has already been completed for the same dataset.
+
+## MVP Specifications
+
+**_Set a cache value_**
+
+To set a cache value, send a POST request with content-type "multipart/form-data" (to facilitate the file upload).
+
+The fields are:
+
+* `service` - required - name of the service
+* `token` - required - authentication token for the service
+* `params` - required - an arbitrary set of data used in the cache's namespace
+* `file` - required - multipart form-data file to cache
+* `expiration_days` - optional - total time in days until the cache's file gets deleted
+
+```
+curl -i -X POST -H "Content-Type: multipart/form-data"
+  -F file=@/home/u/bio.zip
+  -F service=MyService
+  -F token=xyz
+  -F params="{..arbitrary_json_data..}"
+  -F expiration_days=30
+  https://<caching_service_url>/v1/set
+```
+
+If the cache succeeds, you will get back an empty 200 response.
+
+If your auth token is invalid, you will receive a 403. Otherwise, there should be no errors.
+
+**_Request a cache value_**
+
+To fetch a cache value, make a POST request to `/v1/get` with content-type
+"application/x-www-form-urlencoded" and form fields for "service", "token", and "params".
+
+```
+curl -i -X POST -H "Content-Type: application/x-www-form-urlencoded"
+  -F service=MyService
+  -F token=xyz
+  -F params="{..arbitrary_json_data..}"
+  https://<caching_service_url>/v1/get
+```
+
+The service token and params JSON must exactly match those that you used when setting the cache in order to get the same file back.
+
+If the cache file is found, then the server will response with a 200 and the URL of the cached file. Otherwise, the response will have status 404 with an empty body. A 403 response indicates that the service token is invalid. 
+
+### Authentication
```

Comment:
The other use case we discussed was user-specific cached data. There were 2 approaches:

1) Make the services handle that by putting the username in the address of the file somehow
2) Make the caching service handle it by allowing submission of a user token, which would automatically be added to the namespace of the file

Reply:
You can also simply leave this up to the consumer. The consumer can pass a user ID into `params`, which then becomes the cache namespace/identifier

Reply:
That's basically case 1. In that case, every consumer has to duplicate code dealing with this issue. My point is that maybe we should have the service support it and not have that code duplication. Also, if the service supports it users could see what's cached on their behalf across all consumers of the API.